core/src/storage/impls/state_manager.rs

```diff
@@ -261,6 +261,19 @@ impl StateManager {
             // When the delta_height is set to None (e.g. in tests), we
             // assume that the snapshot shift check is
             // disabled.
+            let mut in_progress =
```

Comment:
This write lock prevent other thread from getting an existing snapshot when delta_height == CAPACITY.
It's better to move this code block into the _internal() call. Dive into this logic only if snapshot_db can't be found. Eventually we need a better way of locking, which won't require a write() lock, e.g. CondVar.

Reply:
Actually, you may commit leaving this issue unsolved.